integrationtests/src/test/java/org/axonframework/integrationtests/testsuite/student/EventProcessingAnnotatedStateBasedPooledStreamingTest.java

```diff
@@ -0,0 +1,167 @@
+import org.axonframework.modelling.StateManager;
+import org.axonframework.modelling.configuration.StateBasedEntityModule;
+import org.axonframework.modelling.repository.InMemoryRepository;
+import org.axonframework.serialization.Converter;
+import org.junit.jupiter.api.*;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.testcontainers.shaded.org.awaitility.Awaitility.await;
+
+/**
+ * Test class validating the declarative {@link org.axonframework.eventhandling.pooled.PooledStreamingEventProcessor}
+ * used as a projector for a read model based on {@link StateBasedEntityModule}.
+ *
+ * @author Mateusz Nowak
+ * @since 5.0.0
+ */
+public class EventProcessingAnnotatedStateBasedPooledStreamingTest extends AbstractStudentTestSuite {
+
+    @Test
+    void whenStudentEnrolledThenUpdateReadModel() {
+        // given
+        startApp();
+
+        // when
+        var studentId = UUID.randomUUID().toString();
+        studentEnrolledToCourse(studentId, "my-courseId-1");
+        studentEnrolledToCourse(studentId, "my-courseId-2");
+        studentEnrolledToCourse(studentId, "my-courseId-3");
+
+        // then
+        verifyReadModelState(studentId, state -> {
+            assertThat(state).isNotNull();
+            assertThat(state.courses).containsExactly("my-courseId-1", "my-courseId-2", "my-courseId-3");
+        });
+    }
+
+    record StudentCoursesReadModel(String studentId, List<String> courses) {
+
+        StudentCoursesReadModel(String studentId) {
+            this(studentId, new ArrayList<>());
+        }
+
+        StudentCoursesReadModel evolve(StudentEnrolledEvent event) {
+            var updatedCourses = new ArrayList<>(courses);
+            updatedCourses.add(event.courseId());
+            return new StudentCoursesReadModel(studentId, updatedCourses);
+        }
+    }
+
+    @Override
+    protected EventSourcingConfigurer testSuiteConfigurer(EventSourcingConfigurer configurer) {
+        var studentCoursesRepository = new InMemoryRepository<>(String.class, StudentCoursesReadModel.class);
+        StateBasedEntityModule<String, StudentCoursesReadModel> studentCoursesEntity =
+                StateBasedEntityModule.declarative(String.class, StudentCoursesReadModel.class)
+                                      .repository(cfg -> studentCoursesRepository)
+                                      .messagingModel((configuration, builder) -> builder
+                                              .entityEvolver(readModelEvolver())
+                                              .build())
+                                      .entityIdResolver(cfg -> (message, context) -> {
+                                          var payload = (StudentEnrolledEvent) message.withConvertedPayload(
```

Comment:
I don't think that cast is necessary here.

Reply:
Ah, sorry - I've forgot about the `payloadAs` method :)